### PR TITLE
[CI] Add jshint and stylelint

### PR DESCRIPTION
**Pull request by Frenzie**
This is a bit of a conversation starter. Please take a look at the Travis job output and share which checks you think should be disabled or different. Then it can be adjusted and added as an allowed_fail for the short term or all fixed up immediately, depending on available time.

### CODE REVIEW

**Comment by Spuds** [comment]
This will be an awesome thing to do ... I don't know if this answers your question fully, but I've started to use csscomb to format / cleanup the css.  Having them all "normalized" makes converting my 1.0 themes to 1.1 a bit less of a PITA.  The config that I'm using is as follows:
```
{
    "always-semicolon": true,
    "color-case": "upper",
    "block-indent": "\t",
    "color-shorthand": false,
    "element-case": "lower",
    "eof-newline": true,
    "leading-zero": true,
    "quotes": "double",
    "sort-order-fallback": "abc",
    "space-before-colon": "",
    "space-after-colon": " ",
    "space-before-combinator": " ",
    "space-after-combinator": " ",
    "space-between-declarations": "\n",
    "space-before-opening-brace": " ",
    "space-after-opening-brace": "\n",
    "space-before-selector-delimiter": "",
    "space-before-closing-brace": "\n",
    "strip-spaces": true,
    "unitless-zero": true,
    "vendor-prefix-align": true,
    "sort-order": [
        [
            "font",
            "font-family",
            "font-size",
            "font-weight",
            "font-style",
            "font-variant",
            "font-size-adjust",
            "font-stretch",
            "font-effect",
            "font-emphasize",
            "font-emphasize-position",
            "font-emphasize-style",
            "font-smooth",
            "line-height"
        ],
        [
            "position",
            "z-index",
            "top",
            "right",
            "bottom",
            "left"
        ],
        [
            "display",
            "visibility",
            "float",
            "clear",
            "overflow",
            "overflow-x",
            "overflow-y",
            "-ms-overflow-x",
            "-ms-overflow-y",
            "clip",
            "zoom",
            "-webkit-align-content",
            "-ms-flex-line-pack",
            "align-content",
            "-webkit-box-align",
            "-moz-box-align",
            "-webkit-align-items",
            "align-items",
            "-ms-flex-align",
            "-webkit-align-self",
            "-ms-flex-item-align",
            "-ms-grid-row-align",
            "align-self",
            "-webkit-box-flex",
            "-webkit-flex",
            "-moz-box-flex",
            "-ms-flex",
            "flex",
            "-webkit-flex-flow",
            "-ms-flex-flow",
            "flex-flow",
            "-webkit-flex-basis",
            "-ms-flex-preferred-size",
            "flex-basis",
            "-webkit-box-orient",
            "-webkit-box-direction",
            "-webkit-flex-direction",
            "-moz-box-orient",
            "-moz-box-direction",
            "-ms-flex-direction",
            "flex-direction",
            "-webkit-flex-grow",
            "-ms-flex-positive",
            "flex-grow",
            "-webkit-flex-shrink",
            "-ms-flex-negative",
            "flex-shrink",
            "-webkit-flex-wrap",
            "-ms-flex-wrap",
            "flex-wrap",
            "-webkit-box-pack",
            "-moz-box-pack",
            "-ms-flex-pack",
            "-webkit-justify-content",
            "justify-content",
            "-webkit-box-ordinal-group",
            "-webkit-order",
            "-moz-box-ordinal-group",
            "-ms-flex-order",
            "order"
        ],
        [
            "-webkit-box-sizing",
            "-moz-box-sizing",
            "box-sizing",
            "width",
            "min-width",
            "max-width",
            "height",
            "min-height",
            "max-height",
            "margin",
            "margin-top",
            "margin-right",
            "margin-bottom",
            "margin-left",
            "padding",
            "padding-top",
            "padding-right",
            "padding-bottom",
            "padding-left"
        ],
        [
            "table-layout",
            "empty-cells",
            "caption-side",
            "border-spacing",
            "border-collapse",
            "list-style",
            "list-style-position",
            "list-style-type",
            "list-style-image"
        ],
        [
            "content",
            "quotes",
            "counter-reset",
            "counter-increment",
            "resize",
            "cursor",
            "-webkit-user-select",
            "-moz-user-select",
            "-ms-user-select",
            "user-select",
            "nav-index",
            "nav-up",
            "nav-right",
            "nav-down",
            "nav-left",
            "-webkit-transition",
            "-moz-transition",
            "-ms-transition",
            "-o-transition",
            "transition",
            "-webkit-transition-delay",
            "-moz-transition-delay",
            "-ms-transition-delay",
            "-o-transition-delay",
            "transition-delay",
            "-webkit-transition-timing-function",
            "-moz-transition-timing-function",
            "-ms-transition-timing-function",
            "-o-transition-timing-function",
            "transition-timing-function",
            "-webkit-transition-duration",
            "-moz-transition-duration",
            "-ms-transition-duration",
            "-o-transition-duration",
            "transition-duration",
            "-webkit-transition-property",
            "-moz-transition-property",
            "-ms-transition-property",
            "-o-transition-property",
            "transition-property",
            "-webkit-transform",
            "-moz-transform",
            "-ms-transform",
            "-o-transform",
            "transform",
            "-webkit-transform-origin",
            "-moz-transform-origin",
            "-ms-transform-origin",
            "-o-transform-origin",
            "transform-origin",
            "-webkit-animation",
            "-moz-animation",
            "-ms-animation",
            "-o-animation",
            "animation",
            "-webkit-animation-name",
            "-moz-animation-name",
            "-ms-animation-name",
            "-o-animation-name",
            "animation-name",
            "-webkit-animation-duration",
            "-moz-animation-duration",
            "-ms-animation-duration",
            "-o-animation-duration",
            "animation-duration",
            "-webkit-animation-play-state",
            "-moz-animation-play-state",
            "-ms-animation-play-state",
            "-o-animation-play-state",
            "animation-play-state",
            "-webkit-animation-timing-function",
            "-moz-animation-timing-function",
            "-ms-animation-timing-function",
            "-o-animation-timing-function",
            "animation-timing-function",
            "-webkit-animation-delay",
            "-moz-animation-delay",
            "-ms-animation-delay",
            "-o-animation-delay",
            "animation-delay",
            "-webkit-animation-iteration-count",
            "-moz-animation-iteration-count",
            "-ms-animation-iteration-count",
            "-o-animation-iteration-count",
            "animation-iteration-count",
            "-webkit-animation-direction",
            "-moz-animation-direction",
            "-ms-animation-direction",
            "-o-animation-direction",
            "animation-direction",
            "text-align",
            "-webkit-text-align-last",
            "-moz-text-align-last",
            "-ms-text-align-last",
            "text-align-last",
            "vertical-align",
            "white-space",
            "text-decoration",
            "text-emphasis",
            "text-emphasis-color",
            "text-emphasis-style",
            "text-emphasis-position",
            "text-indent",
            "-ms-text-justify",
            "text-justify",
            "letter-spacing",
            "word-spacing",
            "-ms-writing-mode",
            "text-outline",
            "text-transform",
            "text-wrap",
            "text-overflow",
            "-ms-text-overflow",
            "text-overflow-ellipsis",
            "text-overflow-mode",
            "-ms-word-wrap",
            "word-wrap",
            "word-break",
            "-ms-word-break",
            "-moz-tab-size",
            "-o-tab-size",
            "tab-size",
            "-webkit-hyphens",
            "-moz-hyphens",
            "hyphens",
            "pointer-events"
        ],
        [
            "opacity",
            "filter:progid:DXImageTransform.Microsoft.Alpha(Opacity",
            "-ms-filter:\\'progid:DXImageTransform.Microsoft.Alpha",
            "-ms-interpolation-mode",
            "color",
            "border",
            "border-width",
            "border-style",
            "border-color",
            "border-top",
            "border-top-width",
            "border-top-style",
            "border-top-color",
            "border-right",
            "border-right-width",
            "border-right-style",
            "border-right-color",
            "border-bottom",
            "border-bottom-width",
            "border-bottom-style",
            "border-bottom-color",
            "border-left",
            "border-left-width",
            "border-left-style",
            "border-left-color",
            "-webkit-border-radius",
            "-moz-border-radius",
            "border-radius",
            "-webkit-border-top-left-radius",
            "-moz-border-radius-topleft",
            "border-top-left-radius",
            "-webkit-border-top-right-radius",
            "-moz-border-radius-topright",
            "border-top-right-radius",
            "-webkit-border-bottom-right-radius",
            "-moz-border-radius-bottomright",
            "border-bottom-right-radius",
            "-webkit-border-bottom-left-radius",
            "-moz-border-radius-bottomleft",
            "border-bottom-left-radius",
            "-webkit-border-image",
            "-moz-border-image",
            "-o-border-image",
            "border-image",
            "-webkit-border-image-source",
            "-moz-border-image-source",
            "-o-border-image-source",
            "border-image-source",
            "-webkit-border-image-slice",
            "-moz-border-image-slice",
            "-o-border-image-slice",
            "border-image-slice",
            "-webkit-border-image-width",
            "-moz-border-image-width",
            "-o-border-image-width",
            "border-image-width",
            "-webkit-border-image-outset",
            "-moz-border-image-outset",
            "-o-border-image-outset",
            "border-image-outset",
            "-webkit-border-image-repeat",
            "-moz-border-image-repeat",
            "-o-border-image-repeat",
            "border-image-repeat",
            "outline",
            "outline-width",
            "outline-style",
            "outline-color",
            "outline-offset",
            "background",
            "filter:progid:DXImageTransform.Microsoft.AlphaImageLoader",
            "background-color",
            "background-image",
            "background-repeat",
            "background-attachment",
            "background-position",
            "background-position-x",
            "-ms-background-position-x",
            "background-position-y",
            "-ms-background-position-y",
            "-webkit-background-clip",
            "-moz-background-clip",
            "background-clip",
            "background-origin",
            "-webkit-background-size",
            "-moz-background-size",
            "-o-background-size",
            "background-size",
            "box-decoration-break",
            "-webkit-box-shadow",
            "-moz-box-shadow",
            "box-shadow",
            "filter:progid:DXImageTransform.Microsoft.gradient",
            "-ms-filter:\\'progid:DXImageTransform.Microsoft.gradient",
            "text-shadow"
        ]
    ]
}```

**Comment by Frenzie** [comment]
@Spuds Is it normal that the other jobs are failing? >_>

I think that looks pretty much the same as what I added, except that I opted for lowercase & short for hex colors. (And obviously that order stuff is quite detailed.)

**Comment by Frenzie** [comment]
@Spuds Any thoughts on the dot notation warnings?

```
themes/default/scripts/script_elk.js: line 1254, col 12, ['type'] is better written in dot notation.
```

Dot notation may theoretically be the teeniest tiniest bit quicker but that hardly seems like a reason to call it "better." It's just a stylistic preference. But in spite of the stupidly written error message it'd be silly to disable it if Elk prefers dot notation. ;-)

**Comment by Spuds** [comment]
> @Spuds Is it normal that the other jobs are failing? >_>

Its been an ongoing issue that shows up every 6 months or so ... Its a failure in the web fetch check due to something changing.   I'll get it fixed after this PR (as well as rename the duplicate test names doh!)
 
> Edit: oh btw, stylelint also supports a `--fix` flag since a year or so.

Now that is handy!

> Any thoughts on the dot notation warnings?

As you indicated I think its a bit of a preference thing.   I suppose though that when dealing with objects using the dot notation is cleaner.  I know I did convert many over to .dot  over time but it was a low priority activity.

I don't know if you can go a global ignore on anything ```.min.css``` or .```min.js``` but that would be good.

**Comment by Frenzie** [comment]
@Spuds Are empty blocks automatically removed by the CSS minifier? If so I can disable the check for empty blocks.

File: themes/default/css/_besocial/admin_besocial.css [hunk 1]
```diff
@@ -103,11 +103,6 @@
 	background: #FCFCFC;
 }
 
-.theme_options .content {
-	border-top: 1px solid #CCCCCC;
```
Comment (Frenzie) [inline themes/default/css/_besocial/admin_besocial.css]: To be clear, this whole thing is duplicate, the same borders are defined with the same selectors up above.

**Comment by Frenzie** [comment]
@Spuds Any further comments? The PR is close to finished, see the output in <https://travis-ci.org/elkarte/Elkarte/jobs/567883541>.

**Comment by Spuds** [comment]
Only that it is awesome :D ... Need to merge in the 1.1.6 -> 2.0 branch first.

**Comment by Frenzie** [comment]
Haha, thanks!

**Comment by Spuds** [comment]
@Frenzie could you fix the conflict, it was caused by the 1.1.6 updates that I just merged into 2.0.

**Comment by Frenzie** [comment]
@Spuds Sure thing, not sure I'll get around to it tonight. Btw do you know about the empty blocks? To repeat:

> Are empty blocks automatically removed by the CSS minifier? If so I can disable the check for empty blocks.

If they aren't then I should remove the empty blocks from the child themes.

**Comment by Spuds** [comment]
Cool thanks ... 
Yes I saw the empty blocks warning in the output.  Those were emptied when the "structural" elements were moved back to index.css and out of the variant css.  I think they should probably just be removed.  For a while I thought they might be useful as pointers to what you can easily change, but in redoing my themes I'm not finding them that useful.

**Comment by Frenzie** [comment]
Maybe a couple of decades ago but post-Firebug every browser has pretty decent built-in dev tools. :-)

**Comment by Frenzie** [comment]
Cool! Now there's just a few mostly JSHint complaints to solve to make it a nice green checkmark. :-P (That `this` thing could be serious though.)

**Pull request merged**